ui/vty/app/Glue.hs

```diff
@@ -237,76 +240,100 @@ putCardanoEventToUI UiFace{..} ev =
+uiWalletDatasToTree :: [UiWalletData] -> [UiTree]
+uiWalletDatasToTree = map toTree
   where
-    toTree :: WalletData -> UiTree
-    toTree WalletData {..} =
+    toTree :: UiWalletData -> UiTree
+    toTree UiWalletData {..} =
         Node
-            { rootLabel = UiTreeItem (Just _wdName) [] False
-            , subForest = toList $ map toAccountNode _wdAccounts
+            { rootLabel = UiTreeItem (Just _uwdName) [] False
+            , subForest = toList $ map toAccountNode _uwdAccounts
             }
       where
-        toAccountNode :: AccountData -> UiTree
-        toAccountNode AccountData {..} =
+        toAccountNode :: UiAccountData -> UiTree
+        toAccountNode UiAccountData {..} =
             Node
                 { rootLabel =
                       UiTreeItem
-                          { wtiLabel = Just _adName
-                          , wtiPath = [fromIntegral _adPath]
+                          { wtiLabel = Just _uadName
+                          , wtiPath = [fromIntegral _uadPath]
                           , wtiShowPath = True
                           }
-                , subForest = []
+                , subForest = toList $ map (toAddressNode _uadPath) _uadAddresses
+                }
+        toAddressNode :: Word32 -> (Word32, Address) -> UiTree
+        toAddressNode accIdx (addrIdx, address) =
+            pure $
+            UiTreeItem
+                { wtiLabel = Just (pretty address)
+                , wtiPath = map fromIntegral [accIdx, addrIdx]
+                , wtiShowPath = True
                 }
 
-walletSelectionToPane :: UserSecret -> WalletSelection -> UiWalletInfo
-walletSelectionToPane us WalletSelection{..} = UiWalletInfo{..}
+-- TODO: change to use chain type level
+walletSelectionToPane :: [UiWalletData] -> UiWalletSelection -> UiWalletInfo
+walletSelectionToPane uiwd UiWalletSelection{..} = UiWalletInfo{..}
   where
-    wpiWalletIdx = wsWalletIndex
-    wpiPath = wsPath
-    (wpiType, wpiLabel, wpiAddresses) = case us ^? (usWallets . ix (fromIntegral wsWalletIndex)) of
+    wpiWalletIdx = uwsWalletIdx
+    wpiPath = uwsPath
+    (wpiType, wpiLabel, wpiAddresses) = case uiwd ^? ix (fromIntegral uwsWalletIdx) of
       Nothing -> error "Invalid wallet index"
-      Just WalletData{..} -> case wsPath of
-        [] -> (Just UiWalletInfoWallet, Just _wdName, [])
-        accIdx:_ -> case _wdAccounts ^? ix (fromIntegral accIdx) of
+      Just UiWalletData{..} -> case uwsPath of
+        [] -> (Just UiWalletInfoWallet, Just _uwdName, [])
+        accIdx:_ -> case _uwdAccounts ^? ix (fromIntegral accIdx) of
           Nothing -> error "Invalid account index"
-          Just AccountData{..} ->
-            (Just $ UiWalletInfoAccount [_adPath], Just _adName, toList $ second pretty <$> _adAddresses)
+          Just UiAccountData{..} ->
+            ( Just $ UiWalletInfoAccount [_uadPath]
+            , Just _uadName
+            , map
+              (\(idx, addr) -> (idx, pretty addr))
```

Comment:
This lambda is `second pretty`, as seen in my code to the left. I don't see a reason to change this line at all.